### PR TITLE
Fatoration and Notable Products merge

### PR DESCRIPTION
**Pull request by OsmarHofman**
This pull request contains the following changes:

- Created the Notable Products expert system;

- Added the following rules to the Notable Products expert system:

  - Identification;

  - ApplyCorrectFormula;

  - Power;

  - Multiplication;

  - Sum;

- Created the Fatoration expert system;

- Added the following rules to the multiplication expert system:

  - Identification;

  - CommonFactor;

  - Groupment;

    - GroupmentCommonFactor;

    - GroupmentSumProduct;

  - DifferenceOfTwoSquares;

    - DifferenceOfTwoSquaresExpandedFormula;

    - DifferenceOfTwoSquaresSumDifferenceProduct;

  - PerfectPolynomial;

    - PerfectPolynomialExpandedFormulaConversion;

    - PerfectPolynomialNotableProduct;

  - SecondDegreeTrinomial;

    - SecondDegreeTrinomialConvertToDivisionFormula;

    - SecondDegreeTrinomialProduct;

- Added automated testing for the systems mentioned above;

- Changed the "NumericValueVariable" to "Monomial", since it is a more suitable name;

- Added new methods to the ThreeAddressCode, StringUtil and MathOperatorUtil classes for general purpose;

- Added a new representation of the front-end formulas using HTML entities to help in the view ( see https://dev.w3.org/html5/html-author/charref for the entities used);

- Documentation by JavaDocs pattern.

### CODE REVIEW

**Review by Enedir — APPROVED** [review]

File: src/main/java/br/ifmath/compiler/domain/expertsystem/notableproduct/fatoration/FatorationExpertSystem.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    public void validateExpressions(List<ThreeAddressCode> sources) {
+        List<String> variables = new ArrayList<>();
+        String variable;
+
+        if (StringUtil.isVariable(sources.get(0).getLeft())) {
+            variable = StringUtil.getVariable(sources.get(0).getLeft());
+            NumberUtil.getVariableCoeficient(sources.get(0).getLeft());
+            if (StringUtil.isNotEmpty(variable))
+                variables.add(variable);
+        }
```
Comment (alexjunioravila) [inline src/main/java/br/ifmath/compiler/domain/expertsystem/notableproduct/fatoration/FatorationExpertSystem.java:162]: A lógica desse bloco aparece várias vezes nessa classe (com diferentes parâmetros). Sugiro transformar isso em um método

File: src/main/java/br/ifmath/compiler/infrastructure/util/MathOperatorUtil.java [hunk 2]
```diff
@@ -60,4 +60,9 @@ public static String replaceReducedDistributive(String source) {
         return source;
     }
 
+    public static double gcd(double a, double b) {
```
Comment (alexjunioravila) [inline src/main/java/br/ifmath/compiler/infrastructure/util/MathOperatorUtil.java:63]: O que seria `gcd`? Sugiro colocar documentação explicando isso

Comment (OsmarHofman) [reply src/main/java/br/ifmath/compiler/infrastructure/util/MathOperatorUtil.java:63]: É o método que calcula o maior divisor comum (greatest common divisor) entre dois números. Cheguei a documentar onde esse método é chamado, porém, esqueci de colocar no método. Vou adicionar a documentação.

File: src/main/java/br/ifmath/compiler/domain/expertsystem/notableproduct/fatoration/FatorationExpertSystem.java (hunk 1 above)
Comment (OsmarHofman) [reply src/main/java/br/ifmath/compiler/domain/expertsystem/notableproduct/fatoration/FatorationExpertSystem.java:162]: Farei isso e também adicionarei isso aos outros sistemas especialistas.

**Review by alexjunioravila — APPROVED** [review]

**Pull request merged**